### DISCUSSION TITLE
dynamic and variable tensor shape handling in stan

### DISCUSSION THREAD

**Discussion by dskhanirfan** [stan-dev/cmdstanpy#748] [Q&A] [answered]
0

I have a list of class 'numpy.ndarray', with each tensor of the following possible shapes (all entries will have same shape - each entry can be one of the following possible shapes)

```
Shape of element: ()

Shape of element: (1,)

Shape of element: (128,)

Shape of element: (128, 128, 3, 3, 3)

Shape of element: (128, 256, 1, 1, 1)

Shape of element: (128, 256, 3, 3, 3)

Shape of element: (128, 64, 3, 3, 3)

Shape of element: (256,)

Shape of element: (256, 128, 3, 3, 3)

Shape of element: (256, 256, 3, 3, 3)

Shape of element: (256, 512, 1, 1, 1)

Shape of element: (256, 512, 3, 3, 3)

Shape of element: (32,)

Shape of element: (32, 32, 3, 3, 3)

Shape of element: (32, 4, 3, 3, 3)

Shape of element: (32, 64, 1, 1, 1)

Shape of element: (4,)

Shape of element: (4, 64, 3, 3, 3)

Shape of element: (512,)

Shape of element: (512, 256, 3, 3, 3)

Shape of element: (512, 512, 3, 3, 3)

Shape of element: (64,)

Shape of element: (64, 128, 1, 1, 1)

Shape of element: (64, 128, 3, 3, 3)

Shape of element: (64, 32, 3, 3, 3)

Shape of element: (64, 64, 3, 3, 3)
```
how can I pass it to a dynamic input in stan which can handle this variable type of input dimensions

**Comment by mitzimorris** [comment]
> all entries will have same shape

if all entries have the same shape, then you can specify the number of dimensions and their sizes as data.

in the future, please ask your question on the Stan Forums: https://discourse.mc-stan.org
alternatively, consult the Stan Reference Manual, which describes the Stan language, and the Stan User's Guide: https://mc-stan.org/docs/

**Comment by bob-carpenter** [comment, marked as answer]
> all entries will have same shape - each entry can be one of the following possible shapes
> ...
> how can I pass it to a dynamic input in stan which can handle this variable type of input dimensions

I'm not sure what you mean by all entries having the same shape, since you have listed a bunch of different shapes.

In Stan, you need to specify the shape (e.g., matrix vs. vector vs. row vector) and the size of all variables other than function arguments.  There's no way to write the equivalent of polymorphic Python code for NumPy's `ndarray` within Stan.  For any one of the shapes listed, you can write the corresponding Stan program, but there's no way to do it generically for arguments of different dimensions.  That is, you can write code that works for different size matrices, but you can't have a variable that is sometimes a vector, sometimes a matrix, etc., as you can in Python or R.  

P.S.  As @mitzimorris said, we prefer to have these kinds of questions on the forums rather than as issues, as they're not issues to fix in Stan itself.  Having said that, we're happy to engage wherever users show up.

**Reply by dskhanirfan** [reply]
Thank you. I have a list of tensors (of the same shape) However, in different iterations the list will have different shape of tensors i-e shape of tensors in one iteration will be same. Full description of the modelling problem is here: [https://shorturl.at/efwAC](https://shorturl.at/efwAC)
